packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -49,7 +57,103 @@ export function synthesize(root: IConstruct, options: SynthesisOptions = { }): c
   // stacks to add themselves to the synthesized cloud assembly.
   synthesizeTree(root, builder, options.validateOnSynthesis);
 
-  return builder.buildAssembly();
+  const assembly = builder.buildAssembly();
+
+  invokeValidationPlugins(root, builder.outdir, assembly);
+
+  return assembly;
+}
+
+class LazyHash {
+  private hash?: string;
+
+  constructor(private readonly outdir: string) {}
+
+  get value(): string {
+    if (!this.hash) {
+      this.hash = computeChecksumOfFolder(this.outdir);
+    }
+    return this.hash;
+  }
+}
+
+function invokeValidationPlugins(root: IConstruct, outdir: string, assembly: CloudAssembly) {
+  const lazyHash = new LazyHash(outdir);
+
+  const templatePathsByPlugin: Map<IPolicyValidationPlugin, string[]> = new Map();
+  visit(root, 'post', construct => {
+    if (Stage.isStage(construct)) {
+      for (const plugin of construct.validationPlugins) {
+        if (!templatePathsByPlugin.has(plugin)) {
+          templatePathsByPlugin.set(plugin, []);
+        }
+        templatePathsByPlugin.get(plugin)!.push(...assembly.stacks.map(stack => stack.templateFullPath));
+      }
+    }
+  });
+
+  const reports: NamedValidationPluginReport[] = [];
+  for (const [plugin, paths] of templatePathsByPlugin.entries()) {
+    const originalHash = lazyHash.value;
+
+    try {
+      const report = plugin.validate({ templatePaths: paths });
+      reports.push({ ...report, pluginName: plugin.name });
+    } catch (e) {
+      reports.push({
+        success: false,
+        pluginName: plugin.name,
+        violations: [],
+        metadata: {
+          error: `Validation plugin '${plugin.name}' failed: ${e.message}`,
+        },
+      });
+    }
+    if (computeChecksumOfFolder(outdir) !== originalHash) {
+      throw new Error(`Illegal operation: validation plugin '${plugin.name}' modified the cloud assembly`);
+    }
+  }
+
+  const failed = reports.some(r => !r.success);
+  if (failed) {
+    const tree = new ConstructTree(root);
+    const formatter = new PolicyValidationReportFormatter(tree);
+    const formatJson = FeatureFlags.of(root).isEnabled(cxapi.VALIDATION_REPORT_JSON) ?? false;
+    const output = formatJson
+      ? formatter.formatJson(reports)
+      : formatter.formatPrettyPrinted(reports);
+
+    if (formatJson) {
+      // eslint-disable-next-line no-console
+      console.log(JSON.stringify(output, undefined, 2));
```

Comment:
If the report is printed that means that the CLI will error and all the messages will print to `stderr` while the report will be printed to `stdout` So you can do something like `cdk synth > validations.json` and only the report will be written to the file.

Reply:
> CLI will error and all the messages will print to stderr

Not sure I follow, do we enforce that? what stops us from adding a `console.log` after the validation failed?
Looking at the test:

https://github.com/aws/aws-cdk/blob/4e75df89e7341c7d1a3c6b531f19da5256395d12/packages/%40aws-cdk/core/test/validation/validation.test.ts#L339

This only gets the first call, so it would pass if another call is added, but `stdout` will be mangled. And enforcing it with a test is problematic because it might not cover all code paths in the future.

Reply:
I added an assertion that there is only the single call.

Reply:
I actually think you are correct. I'll write it to a file and we can make the filename configurable later

Reply:
ok this is now written to a file